### ISSUE TITLE
Add temporal upscaling to Sky Shader low res buffers

### ISSUE THREAD

**Issue by clayjohn** [godotengine/godot-proposals#3855]
### Describe the project you are working on

A real time implementation of dynamic, volumetric clouds

### Describe the problem or limitation you are having in your project

Drawing clouds in a shader is really slow and very resource intensive. 

In editor, using half res, on an HD screen with an FX 5600 XT it takes between 4 ms and 11 ms depending on how much of the sky is in view. While this is fine for a tech demo, it is prohibitively expensive for an actual game. 

Rendering clouds at quarter resolution results in too much blurriness if targeting realistic looking clouds.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

This helps overcome the problem as it allows the sky to be rendered at full resolution, but spreads the computation time over 4 or 16 frames. In practice, this means that you can render your sky to the quarter resolution framebuffer and have it look as good as if you rendered it at full resolution. 

In my early tests, the frame rate goes from about 11 ms to about 2 ms. The total cost of rendering the sky thus goes from 10.5 ms to 1.5 ms. 

While 1.5 ms is still a significant portion of a frame budget, it becomes viable for many projects. 

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

 What I am thinking right now is that we would expose two new render_modes: ``upscale_quarter_res_buffer`` and ``upscale_half_res_buffer``. These render modes would then insert the upsampling step into the sky pipeline and ensure that the main pass of the sky shader used the upsampled texture instead of the half res or quarter res texture. 

### If this enhancement will not be used often, can it be worked around with a few lines of script?

This cannot be worked around.

### Is there a reason why this should be core and not an add-on in the asset library?

This cannot be an add-on in the asset library

**Comment by clayjohn** [comment]
I have implemented temporal upscaling in my own branch for people to test! https://github.com/clayjohn/godot/tree/VULKAN-upscale-sky

Running my volumetric cloud demo using this branch brings the frame time down to about 2.5 ms per frame https://github.com/clayjohn/godot-volumetric-cloud-demo

There are still some minor artifacts during fast wind speeds and in the corners during fast camera rotations.

**Comment by clayjohn** [comment]
> Would it make sense to use temporal upscaling for the built-in ProceduralSkyMaterial and PhysicalSkyMaterial? I noticed that using PhysicalSkyMaterial causes a slight increase in GPU frame time (0.1-0.2 ms)[1](#user-content-fn-1-da0a9e39ae90b74df3d1ac600c4d6086) compared to ProceduralSkyMaterial. Perhaps using temporal upscaling could decrease this cost without decreasing quality by rendering at a lower resolution.

Yeah, temporal upscaling is really helpful for any material that changes slowly. For ProceduralSkyMaterial and PhysicalSkyMaterial you wouldn't even notice it unless you suddenly change a property (e.g. if you switch from day to night in a single frame). We would have to expose it as an option to the material for users who do need quick transitions. 

> There's also the option of adding a property to make those skies render at half (or quarter) resolution without temporal upscaling. Since those skies lack high-frequency detail, this could be done without losing much detail.

That's definitely an option. I remember quarter-res looking a little too low resolution when testing (you get banding along the horizon), but half res may be acceptable. 

One thing to keep in mind, when doing low res rendering, you currently don't get to take advantage of the depth buffer. You have to render the whole (low res) buffer each frame. When rendering only high res, you only render the visible pixels. So for games that have lots of objects in view, the benefit of rendering to half res may not be as clear. 

As an aside, I have paused my work on this as JFons is working on TAA/TSS in the main renderer. Once we have TAA/TSS in the main renderer the implementation of TSS for sky shaders will look a little different (I think we can just combine it with the main TAA/TSS).

**Comment by Zylann** [comment]
I have a project with planet volumetric atmospheres you can fly into. Each atmosphere is currently rendered like so:
- When close, a full-screen post-processing quad renders it
- When far, simple geometry covering the planet does the same thing, so it takes a smaller portion of the screen and can be layered with other atmospheres

My goal would be to get the same kind of cloud rendering as seen in No Man Sky or in UE5's default scene.

As mentionned, the problem is how expensive that is. My current prototype just extends atmosphere rendering to raymarch clouds between the outermost sphere to the ground (or other side of the sphere):
![image](https://user-images.githubusercontent.com/1311555/164046135-2ec4018f-6bd9-4834-a9b8-f8dae21b0078.png)
![image](https://user-images.githubusercontent.com/1311555/164046239-396aaa04-1a8c-49ce-acd4-0266fb4b0b27.png)

Because it uses low amount of samples (16 here), it uses random jittering of the ray to get a more coherent result (if not, it produces banding which is quite disturbing). However it produces noise.

To render this at medium and close ranges, it seems indeed important to use temporal techniques + low-res rendering. I can clearly see that's what they do in [NMS](https://www.youtube.com/watch?v=vI8w0l-kLgI).

My concerns are:
- Depth is actually necessary to render such clouds, because you can fly into them, and seeing them through ground, mountains or overlapping close geometry would look bad. Is it really impossible to have access to the depth buffer of the main frame (from the depth prepass?) when doing low-res rendering?
- This proposal targets sky shaders. However that sounds like a restricted pipeline: does that mean I can't fly through clouds rendered in this way? I don't really have a conventional "sky" here, as my worlds are planets I can fly in an out of. I rather have multiple "volumes", without being just fog either.
- Temporal techniques to smooth out noise jittering would have to apply on the low-res render. TAA wouldn't be effective if done on the final result because the noise to smooth out will have been upscaled. Although I guess it depends on the technique?

**Comment by Calinou** [comment]
> One thing to keep in mind, when doing low res rendering, you currently don't get to take advantage of the depth buffer. You have to render the whole (low res) buffer each frame. When rendering only high res, you only render the visible pixels. So for games that have lots of objects in view, the benefit of rendering to half res may not be as clear.

To get this depth prepass optimization back when rendering at half or quarter resolution, can we scale the depth buffer with a shrink operation (to prevent pixels that should be shaded from being considered as occluded by solid geometry)?
